lib/web_ui/lib/src/engine/renderer.dart

```diff
@@ -213,5 +213,5 @@ abstract class Renderer {
 
   ui.ParagraphBuilder createParagraphBuilder(ui.ParagraphStyle style);
 
-  void renderScene(ui.Scene scene);
+  Future<void> renderScene(ui.Scene scene);
```

Comment:
This is being called from a synchronous method:
https://github.com/flutter/engine/blob/2e5cdeade5cc15e52589099f571347987b5fb3ba/lib/web_ui/lib/src/engine/platform_dispatcher.dart#L708-L710

I don't think we should change the signature of `Renderer.renderScene`. Since `skwasm` is the only one that needs the asynchronicity, let's only change `SkwasmRenderer.renderScene`'s signature? And I would argue that even `SkwasmRenderer.renderScene` doesn't need to be asynchronous. Nothing is awaiting it, so effectively, we are firing the future and forgetting about it.

One concern that I have here though is race conditions. E.g. what if the previous `renderScene` call takes a while and we receive a new one? Should we somehow force these calls to go in sequence (by pushing them into a queue for example)? Again, if this is something we plan to handle later, let's add a TODO.

Reply:
It is fine for this to be called and not awaited from a synchronous method in a "fire and forget" manner, and that's what is still going on right now. Making this method async at the renderer level only makes things a bit more expressive and flexible. The existing renderers which do not asynchronously render can return an immediately resolved future.

The main impetus behind doing this is because the new golden screenshot capabilities in the golden tests need to wait for the rendering to be actually complete before attempting a screenshot. So really, this future is only being awaited in tests.

As for your concern in the last paragraph, I have a lot of the same thoughts. That's sort of the reason why I'd like to expose this as an async method. In situations where the render thread is slower than our target FPS, we may want to keep track of pending futures and stop actually submitting new frames if it starts getting backed up. Otherwise, if we continue pushing more frames, the render thread will get continually more backed up, when it's better to to just drop frames. We can't do this with our current "fire-and-forget" method, but making the method async opens us up to doing this in the future.

Reply:
Agreed with Mouad, this should be synchronous for consistency with the mobile Flutter engine and our public facing API. The Skwasm renderer can have a way to `await` for the scene to actually be rendered which can be used in tests. As for the problem with the render thread getting backed up, this is already something that can happen in Flutter mobile and I think they just drop everything but the latest frame.

Reply:
Is there a suggested alternative? This `Renderer.renderScene` API is web-specific, so there is nothing in the mobile Flutter engine to keep it consistent with, and I'm also not changing the public facing API (`PlatformDispatcher.render` is still a synchronous method). Basically the reason I want to return a Future here is:
* To allow us to wait for the render to complete in the tests.
* If we do want to "drop everything but the latest frame" like the native engine does, the UI thread still needs a way to tell whether we are actually getting backed up.
A `Future` returned directly from the method seems like the most natural way to achieve this, but if there is something different you all have in mind I'm open to suggestions.

Reply:
Maybe I could have `Renderer.renderScene` return `FutureOr<void>`, so that the subclasses of `Renderer` that aren't asynchronous can still just return `void`?

Reply:
One way is to only make `SkwasmRenderer.renderScene` asynchronous, and leave others synchronous:
```dart
abstract class Renderer {
  void renderScene();
}

class OtherRenderer implements Renderer {
  @override
  void renderScene() {
    print('foo');
  }
}

class SkwasmRenderer implements Renderer {
  // `Future<void Function()` can override `void Function()`.
  @override
  Future<void> renderScene() async {
    await Future.delayed(Duration(milliseconds: 2));
    print('foo');
  }
}
```
Another way is to keep `SkwasmRenderer.renderScene` synchronous too, and add a separate async method:
```dart
abstract class Renderer {
  void renderScene();
}

class OtherRenderer implements Renderer {
  @override
  void renderScene() {
    print('foo');
  }
}

class SkwasmRenderer implements Renderer {
  @override
  void renderScene() {
    // Fire and forget.
    renderSceneAndWait();
  }
  
  @visibleForTesting
  Future<void> renderSceneAndWait() async {
    await Future.delayed(Duration(milliseconds: 2));
    print('foo');
  }
}
```

Both of these approaches will be annoying if you want a test that runs against all renderers:
```dart
for (final Renderer renderer in renderers) {
  // In order to access the `renderSceneAndWait` method, you need to do:
  if (renderer is SkwasmRenderer) {
    await renderer.renderSceneAndWait();
  } else {
    renderer.renderScene();
  }
}
```

Reply:
I've already switched over to `FutureOr<void>`, which allows some renderers to be sync and others to be async. I don't understand why we would want to add the complexity of having a layer that does "fire and forget" within the skwasm renderer, when we could simply "fire and forget" from the `PlatformDispatcher`, which is what we are already doing. It's worth mentioning that my change as written makes no change to production behavior, and does not modify the public API. I don't really understand the objection, and why we would want to add the complexity of separate `renderScene` vs `renderSceneAndWait` APIs and class introspection when the behavior can be sufficiently encapsulated generically with a single method.

Reply:
I should've clarified my motivation for keeping things as `sync` as possible. It's because I want future maintainers to be aware that this is used as a sync API and it's better to stay that way. And if they really need the asynchronicity, then they can make the deliberate decision of converting it to async.

If we make all of them async today, I'm afraid it becomes too easy in the future to add async stuff in there without thinking about the consequences. I want to conversion to async to be a deliberate decision based on a real need (like in the case of skwasm).

Reply:
Does the `FutureOr<void>` change not address your issue? `SkwasmRenderer`'s render will return `Future<void>`, but the other renderers still return `void`. So it would require us to explicitly change a renderer's return value to make it async

Reply:
It doesn't address the issue. The developer making change to `renderScene` will think it's okay to add an `await` in there because the signature already "supports" async. I want them to have the extra burden of changing the signature, which will make them look around and check whether it's fine to convert it to async or not.

That said, I don't feel strongly about it, and maybe this is all just me doing premature design optimizations :)

Reply:
Just to clarify, if a developer is editing the `renderScene` function on one of the existing renderers (whose `renderScene` method is synchronous), and they want to add an await, they *will* have the extra burden of changing the signature, since the signature of the existing renderers is staying `void renderScene(...)`. They would have to explicitly change it to `Future<void> renderScene(...)`. The abstract `Renderer` class has the signature `FutureOr<void> renderScene(...)`, but the actual classes themselves either explicitly return `void` or explicitly return `Future<void>`

Reply:
Ahh okay got it now. I misunderstood your previous comment, thought you were making subclasses `FutureOr<void>` too.

Ok, this actually addresses my concerns, thanks!